File: src/process.zig
const kernel = @import("kernel.zig");
const std = @import("std.zig");
const zstd = @import("std");
const sbi = @import("sbi.zig");
const PROCS_MAX = 8;
const KERNEL_STACK_SIZE = 1024 * 20;
const PROC_UNUSED = false;
const PROC_RUNNABLE = true;

pub extern fn switch_context(prev_sp: *usize, next_sp: *usize) noreturn;

pub const Process = struct {
    pid: usize,
    state: bool,
    sp: usize,
    stack: [KERNEL_STACK_SIZE]u8 align(16),
};

var procs: [PROCS_MAX]Process = .{Process{
    .pid = undefined,
    .state = PROC_UNUSED,
    .sp = undefined,
    .stack = [_]u8{0} ** KERNEL_STACK_SIZE,
}} ** PROCS_MAX;

pub fn create(pc: usize) *Process {
    std.print("criando processo com pc = {x}\n", .{pc});
    var proc: ?*Process = null;

    var pid: usize = 0;

    // Encontrar um slot livre para esse processo
    for (0..procs.len) |i| {
        if (procs[i].state == PROC_UNUSED) {
            std.print("encontramos um slot para o processo em i={d}\n", .{i});
            std.print("proc = {*}\n", .{&(procs[i])});
            proc = (&(procs[i]));
            pid = i;
            break;
        }
    }
    std.print("O pid para o processo {x} é {d}\n", .{ pc, pid });

    if (proc == null) {
        kernel.PANIC("Não há mais slots livres de memória", .{}, @src());
    }

    var sp: [*]usize = blk: {
        const stack_cast: [*]u8 = @ptrCast(&proc.?.stack); // Para apontarmos
        // pro topo precisamos
        // que ele seja um many-item
        // pointer
        const stack_top = stack_cast + (proc.?.stack.len);
        break :blk @alignCast(@ptrCast(stack_top));
    };
    std.print("sp = {*}\n", .{sp});

    var i: usize = 12;
    while (i > 0) : (i -= 1) {
        sp -= 1;
        const ptr: *usize = @ptrCast(sp);
        ptr.* = 0;
    }
    var ptr: *u32 = @ptrCast(sp);
    sp -= 1;
    ptr = @ptrCast(sp);
    ptr.* = pc;

    proc.?.pid = pid;
    proc.?.state = PROC_RUNNABLE;
    proc.?.sp = @intFromPtr(sp);
    std.print("sp = {*}, proc.sp = {x}\n", .{ sp, proc.?.sp });

    const stack_u32: []u32 = @ptrCast(@alignCast(&(proc.?.stack)));
    std.print("stack: {x}\n", .{stack_u32[stack_u32.len - 13 ..]});
    std.print("processo: {*}\n", .{proc.?});
    std.print("processo: {*}\n", .{&(procs[pid])});

    for (procs) |value| {
        std.print("{any}\n", .{value.state});
    }

    return proc.?;
}

var current_proc: *Process = undefined;
var idle_proc: *Process = undefined;

fn teste() void {
    std.print("estamos no teste\n", .{});
}

pub fn init() void {
    idle_proc = create(@intFromPtr(&teste));
    idle_proc.pid = 0;
    current_proc = idle_proc;
}

pub fn yield() void {
    // std.print("yielding\n", .{});
    var next = idle_proc;
    var i: usize = 0;
    while (i < PROCS_MAX) : (i += 1) {
        const idx = (current_proc.pid + i + 1) % PROCS_MAX;
        const proc = &(procs[idx]);
        {
            const idx_u8: u8 = @intCast(idx);
            sbi.putchar('0' + idx_u8);
            sbi.putchar('\n');
        }
        // std.print("proc.pid = {d}\n", .{proc.pid});
        if (proc.state == PROC_RUNNABLE and idx > 0) {
            std.print("proc id = {x}\n", .{proc.pid});
            next = @constCast(proc);
            break;
        }
    }

    // for (procs) |proc| {
    // std.print("proc: {any}\n", .{proc.state});
    // }

    // Se só tiver este único processo
    if (next == current_proc) {
        // std.print("next = {*}, current_proc = {*}\n", .{ next, current_proc });
        return;
    }

    // Mudamos de contexto
    const prev = current_proc;
    current_proc = next;
    // asm volatile (
    //     \\ addi sp, sp, -13*4 # o copilot falou pra eu tomar cuidado com o alinhamento
    //     \\ sw ra, 0 * 4(sp)
    //     \\ sw s0, 1 * 4(sp)
    //     \\ sw s1, 2 * 4(sp)
    //     \\ sw s2, 3 * 4(sp)
    //     \\ sw s3, 4 * 4(sp)
    //     \\ sw s4, 5 * 4(sp)
    //     \\ sw s5, 6 * 4(sp)
    //     \\ sw s6, 7 * 4(sp)
    //     \\ sw s7, 8 * 4(sp)
    //     \\ sw s8, 9 * 4(sp)
    //     \\ sw s9, 10 * 4(sp)
    //     \\ sw s10, 11 * 4(sp)
    //     \\ sw s11, 12 * 4(sp)
    //     \\ 
    //     \\ sw sp, (%[prev_sp])
    //     \\ lw sp, (%[next_sp])
    //     \\
    //     \\ lw ra, 0 * 4(sp)
    //     \\ lw s0, 1 * 4(sp)
    //     \\ lw s1, 2 * 4(sp)
    //     \\ lw s2, 3 * 4(sp)
    //     \\ lw s3, 4 * 4(sp)
    //     \\ lw s4, 5 * 4(sp)
    //     \\ lw s5, 6 * 4(sp)
    //     \\ lw s6, 7 * 4(sp)
    //     \\ lw s7, 8 * 4(sp)
    //     \\ lw s8, 9 * 4(sp)
    //     \\ lw s9, 10 * 4(sp)
    //     \\ lw s10, 11 * 4(sp)
    //     \\ lw s11, 12 * 4(sp)
    //     \\ 
    //     \\ addi sp, sp, 13*4
    //     \\ ret
    //     :
    //     : [prev_sp] "r" (&prev.sp),
    //       [next_sp] "r" (&next.sp),
    // );
    std.print("mudando o contexto\n", .{});
    switch_context(&prev.sp, &next.sp);
}

File: src/process.c
#include "process.h"

// NOTE: FUNCIONAMENTO
// cada processo e é representado pelo strcut process.
// Cada processo possui um pid, um state (RUNNABLE ou UNUSED), um stack pointer
// que aponta para o stack do próprio processo, um stack próprio e uma
// page_table, e cada processo possui a sua própria. A tradução do endereço
// de memória virtual para o endereço de memória real é feito pelo próprio
// processador riscv32.

struct process procs[PROCS_MAX] = {(struct process){
    .pid = 0,
    .state = PROC_UNUSED,
    .sp = 0,
    .page_table = NULL,
    .stack = {0},
}};

// ---- Funcionamento ----
// Para criarmos um processo, primeiro precisamos descorbrir um espaço livre
// na nossa array de processos `procs`, isso vai determinar o pid do nosso
// processo e o ponteiro para ele na array de processos.
// Em seguida preenchemos o stack com os valores dos callee saved registers
// (s0-s11) que inicialmente são zero, e depois no topo do stack guardamos o 
// ndereço do retorno (que no caso é o início da função da qual criamos o pro
// esso).
// TODO: escrever sobre a page table
//
// E por fim retornamos o ponteiro para esse processo (que está armazenado na
// nossa array de processos
// -----------------------
struct process *create_process(uint32_t pc) {
  printf("create_process: pc=%x\n", pc);
  struct process *proc = (struct process *)NULL;
  size_t pid;
  for (pid = 0; pid < PROCS_MAX; pid++) {
    if (procs[pid].state == PROC_UNUSED) {
      proc = &procs[pid];
      break;
    }
  }

  if (proc == NULL) {
    PANIC("Não há mais slots livres");
  }

  uint32_t *sp = (uint32_t *)&proc->stack[PROC_STACK_SIZE];
  for (size_t i = 1; i < 13; i++) { // Preencher o s0 a s11 com zeros
    sp -= 1;
    *sp = 0;
  }
  sp -= 1;
  *sp = (uint32_t)pc; // Preenchendo o ra com o ponteiro da função

  // Por fim fazemos o map das kernel pages
  uint32_t *page_table = (uint32_t *)alloc_pages(1);
  for (paddr_t paddr = (paddr_t)__kernel_base; paddr < (paddr_t)__free_ram_end;
       paddr += PAGE_SIZE) {
    map_page(page_table, paddr, paddr, PAGE_R | PAGE_W | PAGE_X);
  }

  proc->pid = pid + 1;
  proc->state = PROC_RUNNABLE;
  proc->sp = (uint32_t)sp;
  proc->page_table = page_table;
  printf("Processo criado com pid=%x e sp=%x\n", proc->pid, proc->sp);
  return proc;
}

struct process *current_proc = (struct process *)NULL;
struct process *idle_proc = (struct process *)NULL;

// ---- Funcionamento ----
// Não há ordem de prioridade para a troca dos processos, pegamos o processo
// atual e seguimos pela array de processos a partir dele até encontrarmos um
// processo que pode ser rodado. Dessa forma é garantido que não há starvation
// nem deadlock (pelo menos não por parte do yield) pois todos os processos são
// executados uma vez que o processo precedente abre mão do controle
// -----------------------
void yield(void) {
  printf("yelding\n");
  struct process *next = idle_proc;
  for (size_t i = 0; i < PROCS_MAX; i++) {
    struct process *proc = &procs[(current_proc->pid + i) % PROCS_MAX];
    if (proc->state == PROC_RUNNABLE && proc->pid > 0) {
      next = proc;
      break;
    }
  }

  if (next == current_proc) {
    return;
  }

  struct process *prev = current_proc;
  current_proc = next;
  printf("switching context %d -> %d\n", prev->pid, next->pid);

  // NOTE: entender o que está acontecendo aqui
  __asm__ __volatile__(
      "sfence.vma\n"
      "csrw satp, %[satp]\n"
      "sfence.vma\n"
      "csrw sscratch, %[sscratch]\n" // Isso aqui é para o handler de interrupções
                                     // TODO: entender o porque dele
      :
      : [satp] "r"(SATP_SV32 | ((uint32_t)next->page_table / PAGE_SIZE)),
        [sscratch] "r"((uint32_t)&next->stack[sizeof(next->stack)]));

  switch_context(&prev->sp, &next->sp);
}
